### ISSUE TITLE
Are future rewards especially those at the end of an episode taken into account properly?

### ISSUE THREAD

**Issue by tobi45** [tensorflow/agents#745]
I am using a `DynamicEpisodeDriver` to collect data for training and store it in a `TFUniformReplayBuffer` to train a `DqnAgent`. At the end of an episode, the environment gives the agent a final reward with the last step, indicating whether the agent had success or not. The sources are mostly the same as in your dqn agent tutorial but with another environment and an episode driver.

As I digged a bit deeper into the machinery it appears to me that future rewards are not taken into account appropriately.

- The steps collected in the replay buffer have a `reward` that was assigned by the environment and a `discount` factor assigned by environment or gym environment wrapper. Thus up to this point no future rewards are considered.
- During training iterations a number of steps (in my case the default `2` as in the tutorial) are sampled from the dataset backed by the replay buffer using its `as_dataset` method are passed as experience to the `train` method of the agent.
- As mentioned in #503 the agent uses its `gamma` value together with the `discount` of the steps to take future rewards into account

And here is my problem: how is the reward given at the end of an episode taken into account for previous steps? I just see two possibilities:

1. The episode driver has to calculate them before adding the steps into the replay buffer. But that does not happen. Thus, when 2 steps are sampled for training, the agent is only able to take the rewards of those steps into account, because it has no access to other steps. Thus with episodes lasting around `200` steps most of the time the agent is unable to take the reward given in the last episode step into account.
2. A whole episode with all of its steps has to be sampled from the dataset during the training iterations and needs to be passed to the agent. By doing so, the agent should be able to take the reward given by an environment at the end of an episode into account. But to do so, there is no API to fetch a whole episode from the dataset or replay buffer. Furthermore the number of steps within an episode is not fixed but the dqn agent expects a fixed number of steps in its experience for training set with its `n_step_update` init argument.

Here are my questions:

- Do I miss something, conceptually or API or at all?
- Are there some workarounds available?

This issue might be related to: #613

**Comment by sguada** [comment]
So if your environment only provides returns at the last step, then only the last transition would have a return that it's non zero during learning the model would bootstrap and propagate the q-values, but  DQN struggles with tasks that have very sparse rewards. In those cases using n_step_update greater than 2 can help, or you can try a different algorithm, like PPO.

**Comment by tobi45** [comment]
@sguada thank you for you quick answer and the hint to `PPO`!

Indeed I currently experiment with an environment that only provides returns at the last step in an episode. In a probably more realistic setting the agent may get a little negative reward for each step (to learn to not waste time) instead of zero. But the main reward at the last step would remain (task solved or not). Even in such settings the agent should get credit for each step that is done to eventually solve the task (and not stuck with the little negative reward given by the environment). Such credits couldn't be given by the environment as it does not know, what the agent does in the next step. As far as I understand the literature the discount factor is used to take future rewards into account.

As I understand your answer tensorflow agent currently might not support future rewards? During research I discovered an operation that might does this calculation (I haven't evaluated yet): https://www.tensorflow.org/agents/api_docs/python/tf_agents/utils/value_ops/discounted_return  It seems to be used by some agents, but not the dqn.

**Comment by sguada** [comment]
The way DQN works is that it would learn the target q-values and it would propagate them through as it learns, so yeah the discount factor it's used to modulate how much future rewards impact current actions.

Take a look at q-learning https://en.wikipedia.org/wiki/Q-learning

What I meant is that DQN can struggle to learn from very sparse rewards, but no that it doesn't take into account future rewards.

**Comment by tobi45** [comment]
Thank you very much for your answer and clarification! I realized myself the terms reward and return got a bit mixed up to me.

You're right, it's my problem setting, where the reward function and the dqn agent don't play well together. So either I find a reward function that is more suited for the dqn agent or I try other agents.

Thanks a lot for your time, hints and help!
